File: docker/backend/presets.js
'use strict';

// ============================================================
//  presets.js – Betriebsmodi / Preset-Verwaltung
// ============================================================

const fs    = require('fs').promises;
const path  = require('path');
const state = require('./state');
const pi    = require('./pressureCtrl');
const mqtt  = require('./mqttClient');

const DATA_FILE = process.env.DATA_DIR
  ? path.join(process.env.DATA_DIR, 'presets.json')
  : '/data/presets.json';

const DEFAULT_PRESET = {
  name:     'Normal',
  mode:     0,      // 0=Druck, 1=Durchfluss, 2=FixFrequenz
  setpoint: 3.0,
  kp:       8.0,
  ki:       1.0,
  freq_min: 35.0,
  freq_max: 52.0,
  setpoint_hz:       0,    // nur mode=2: feste Frequenz
  expected_pressure: 0,    // nur mode=2: erwarteter Druck (für Trockenlauf-/Überdruck-Schutz)
};

let presets = [{ ...DEFAULT_PRESET }];
let _onChanged = null;
function onPresetsChanged(cb) { _onChanged = cb; }
function _notifyChanged() { if (_onChanged) try { _onChanged(); } catch {} }

function clampHz(v) {
  if (!Number.isFinite(v) || v <= 0) return 0;
  return Math.max(10, Math.min(60, v));
}
function clampPressure(v) {
  if (!Number.isFinite(v) || v <= 0) return 0;
  return Math.max(0.1, Math.min(8.0, v));
}

async function load() {
  try {
    const data = await fs.readFile(DATA_FILE, 'utf8');
    const saved = JSON.parse(data);
    if (Array.isArray(saved) && saved.length > 0) {
      presets = saved;
    }
    // Sicherstellen dass Normal-Preset existiert
    if (!presets.find(p => p.name === 'Normal')) {
      presets.unshift({ ...DEFAULT_PRESET });
    }
    console.log('[Presets] Geladen:', presets.map(p => p.name).join(', '));
  } catch {
    console.log('[Presets] Keine Datei gefunden – Standard-Preset genutzt');
  }
}

async function saveToDisk() {
  await fs.writeFile(DATA_FILE, JSON.stringify(presets, null, 2));
}

function list() {
  return {
    active:  state.active_preset,
    presets: presets.map(p => ({ ...p })),
  };
}

function addOrUpdate(preset) {
  if (!preset.name || preset.name.length > 32) return false;
  const idx = presets.findIndex(p => p.name === preset.name);
  const mode = parseInt(preset.mode);
  const entry = {
    name:     preset.name,
    mode:     (mode === 1 || mode === 2) ? mode : 0,
    setpoint: parseFloat(preset.setpoint) || 3.0,
    kp:       parseFloat(preset.kp)       || 8.0,
    ki:       parseFloat(preset.ki)       || 1.0,
    freq_min: parseFloat(preset.freq_min) || 35.0,
    freq_max: parseFloat(preset.freq_max) || 52.0,
    setpoint_hz:       clampHz(parseFloat(preset.setpoint_hz)),
    expected_pressure: clampPressure(parseFloat(preset.expected_pressure)),
  };
  if (idx >= 0) {
    presets[idx] = entry;
  } else {
    if (presets.length >= 20) return false;
    presets.push(entry);
  }
  saveToDisk().catch(e => console.error('[Presets] save error:', e.message));
  _notifyChanged();
  return true;
}

function deletePreset(name) {
  if (state.active_preset === name) return false;
  const idx = presets.findIndex(p => p.name === name);
  if (idx < 0) return false;
  presets.splice(idx, 1);
  saveToDisk().catch(e => console.error('[Presets] save error:', e.message));
  _notifyChanged();
  return true;
}

function apply(name) {
  const preset = presets.find(p => p.name === name);
  if (!preset) return false;

  pi.resetIntegral();

  if (preset.mode === 2) {
    // Fix-Frequenz-Preset: PI deaktivieren, Frequenz direkt setzen
    state.pi.enabled = false;
    state.pi.ctrl_mode      = 2;
    state.pi.flow_setpoint  = 0;
    state.active_preset     = name;
    state.ctrl_mode         = 2;
    state.preset_expected_pressure = preset.expected_pressure || 0;
    state.preset_setpoint_hz       = preset.setpoint_hz || 0;
    // Frequenz und Start an V20 senden — pressureCtrl.tick() refresht regelmäßig
    if (preset.setpoint_hz > 0) {
      mqtt.sendCmd('v20/start', '1');
      mqtt.sendCmd('v20/freq', preset.setpoint_hz.toFixed(1));
      state.v20.freq_setpoint = preset.setpoint_hz;
    }
    console.log('[Presets] Aktiviert (FixHz):', name, preset.setpoint_hz, 'Hz');
    return true;
  }

  // mode 0/1: PI re-konfigurieren und aktivieren
  // Wenn von Fix-Frequenz-Modus gewechselt wird: Pumpe sofort stoppen
  if (state.ctrl_mode === 2) {
    mqtt.sendCmd('v20/stop', '1');
  }
  pi.setConfig({
    enabled:  true,
    setpoint: preset.setpoint,
    kp:       preset.kp,
    ki:       preset.ki,
    freq_min: preset.freq_min,
    freq_max: preset.freq_max,
  });

  state.pi.ctrl_mode      = preset.mode;
  state.pi.flow_setpoint  = preset.mode === 1 ? preset.setpoint : 0;
  state.active_preset     = name;
  state.ctrl_mode         = preset.mode;
  state.preset_expected_pressure = 0;
  state.preset_setpoint_hz       = 0;

  console.log('[Presets] Aktiviert:', name);
  return true;
}

module.exports = { load, list, addOrUpdate, deletePreset, apply, onPresetsChanged };

File: pi/backend/app/presets.py
"""Betriebsmodi / Preset-Verwaltung — Port von docker/backend/presets.js.

Modi:
  0 = Druck       (PI-Druckregelung)
  1 = Durchfluss  (PI auf flow_setpoint)
  2 = FixHz       (Festfrequenz, PI deaktiviert; Trockenlauf-/Überdruckschutz aktiv)
"""
from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass, field
from typing import Any

from .persistence import PRESETS_FILE, load_json, save_json
from .state import app_state, web_log

MAX_PRESETS = 20


@dataclass
class Preset:
    name: str
    mode: int = 0
    setpoint: float = 3.0
    kp: float = 8.0
    ki: float = 1.0
    freq_min: float = 35.0
    freq_max: float = 52.0
    setpoint_hz: float = 0.0
    expected_pressure: float = 0.0

    def to_dict(self) -> dict[str, Any]:
        return {
            "name": self.name,
            "mode": self.mode,
            "setpoint": self.setpoint,
            "kp": self.kp,
            "ki": self.ki,
            "freq_min": self.freq_min,
            "freq_max": self.freq_max,
            "setpoint_hz": self.setpoint_hz,
            "expected_pressure": self.expected_pressure,
        }


DEFAULT_PRESET = Preset(name="Normal")


def _clamp_hz(v: float) -> float:
    if not isinstance(v, (int, float)) or v <= 0:
        return 0.0
    return max(10.0, min(60.0, float(v)))


def _clamp_pressure(v: float) -> float:
    if not isinstance(v, (int, float)) or v <= 0:
        return 0.0
    return max(0.1, min(8.0, float(v)))


@dataclass
class PresetManager:
    """Hält die Liste der Presets und wendet sie an.

    Abhängigkeiten werden injiziert (pressure_ctrl, V20-Callbacks), damit
    die Logik nicht hart an Modbus-Implementierung gekoppelt ist.
    """
    pi_ctrl: Any                              # PressureController-Instanz
    on_v20_start: Callable[[], None]
    on_v20_freq: Callable[[float], None]

    presets: list[Preset] = field(default_factory=lambda: [Preset(name="Normal")])
    on_changed: Callable[[], None] | None = None

    def load(self) -> None:
        data = load_json(PRESETS_FILE)
        if isinstance(data, list) and data:
            self.presets = [self._from_dict(p) for p in data]
        # Sicherstellen, dass "Normal" existiert
        if not any(p.name == "Normal" for p in self.presets):
            self.presets.insert(0, Preset(name="Normal"))
        web_log(f"[Presets] Geladen: {', '.join(p.name for p in self.presets)}")

    def save(self) -> None:
        save_json(PRESETS_FILE, [p.to_dict() for p in self.presets])

    def list(self) -> dict[str, Any]:
        return {
            "active": app_state.active_preset,
            "presets": [p.to_dict() for p in self.presets],
        }

    def add_or_update(self, body: dict[str, Any]) -> bool:
        name = (body.get("name") or "").strip()
        if not name or len(name) > 32:
            return False
        entry = self._from_dict(body)
        for i, p in enumerate(self.presets):
            if p.name == name:
                self.presets[i] = entry
                self.save()
                self._notify()
                return True
        if len(self.presets) >= MAX_PRESETS:
            return False
        self.presets.append(entry)
        self.save()
        self._notify()
        return True

    def delete(self, name: str) -> bool:
        if app_state.active_preset == name:
            return False
        idx = next((i for i, p in enumerate(self.presets) if p.name == name), -1)
        if idx < 0:
            return False
        self.presets.pop(idx)
        self.save()
        self._notify()
        return True

    def apply(self, name: str) -> bool:
        preset = next((p for p in self.presets if p.name == name), None)
        if preset is None:
            return False

        self.pi_ctrl.set_manual_stop(False)
        self.pi_ctrl._reset_integral()  # private API – bewusst, wie pi.resetIntegral()

        if preset.mode == 2:
            # Fix-Frequenz: PI deaktivieren, Sollfrequenz direkt setzen
            app_state.pi.enabled = False
            app_state.pi.ctrl_mode = 2
            app_state.pi.flow_setpoint = 0
            app_state.active_preset = name
            app_state.ctrl_mode = 2
            app_state.preset_expected_pressure = preset.expected_pressure or 0.0
            app_state.preset_setpoint_hz = preset.setpoint_hz or 0.0
            if preset.setpoint_hz > 0:
                self.on_v20_start()
                self.on_v20_freq(preset.setpoint_hz)
                app_state.v20.freq_setpoint = preset.setpoint_hz
            web_log(f"[Presets] Aktiviert (FixHz): {name} {preset.setpoint_hz} Hz")
            return True

        # Mode 0/1: PI re-konfigurieren
        if app_state.ctrl_mode == 2:
            self.pi_ctrl.force_stop()
        self.pi_ctrl.set_config({
            "enabled": True,
            "setpoint": preset.setpoint,
            "kp": preset.kp,
            "ki": preset.ki,
            "freq_min": preset.freq_min,
            "freq_max": preset.freq_max,
        })
        app_state.pi.ctrl_mode = preset.mode
        app_state.pi.flow_setpoint = preset.setpoint if preset.mode == 1 else 0
        app_state.active_preset = name
        app_state.ctrl_mode = preset.mode
        app_state.preset_expected_pressure = 0
        app_state.preset_setpoint_hz = 0
        web_log(f"[Presets] Aktiviert: {name}")
        return True

    # ── intern ────────────────────────────────────────────────
    def _notify(self) -> None:
        if self.on_changed:
            try:
                self.on_changed()
            except Exception as exc:
                print(f"[Presets] on_changed error: {exc}", flush=True)

    @staticmethod
    def _from_dict(d: dict[str, Any]) -> Preset:
        try:
            mode = int(d.get("mode", 0))
        except (TypeError, ValueError):
            mode = 0
        if mode not in (0, 1, 2):
            mode = 0

        def f(key: str, default: float) -> float:
            try:
                return float(d.get(key, default))
            except (TypeError, ValueError):
                return default

        return Preset(
            name=str(d.get("name", "Unbenannt")),
            mode=mode,
            setpoint=f("setpoint", 3.0),
            kp=f("kp", 8.0),
            ki=f("ki", 1.0),
            freq_min=f("freq_min", 35.0),
            freq_max=f("freq_max", 52.0),
            setpoint_hz=_clamp_hz(f("setpoint_hz", 0.0)),
            expected_pressure=_clamp_pressure(f("expected_pressure", 0.0)),
        )
